moveit_kinematics/test/test_kinematics_plugin.cpp

```diff
@@ -0,0 +1,737 @@
+    // clang-format on
+  }
+  testing::AssertionResult isNear(const char* expr1, const char* expr2, const char* abs_error_expr,
+                                  const geometry_msgs::Quaternion& val1, const geometry_msgs::Quaternion& val2,
+                                  double abs_error)
+  {
+    if ((std::fabs(val1.x - val2.x) <= abs_error && std::fabs(val1.y - val2.y) <= abs_error &&
+         std::fabs(val1.z - val2.z) <= abs_error && std::fabs(val1.w - val2.w) <= abs_error) ||
+        (std::fabs(val1.x + val2.x) <= abs_error && std::fabs(val1.y + val2.y) <= abs_error &&
+         std::fabs(val1.z + val2.z) <= abs_error && std::fabs(val1.w + val2.w) <= abs_error))
+      return testing::AssertionSuccess();
+
+    // clang-format off
+    return testing::AssertionFailure()
+        << std::setprecision(std::numeric_limits<double>::digits10 + 2)
+        << "Expected: " << expr1 << " [" << val1.w << ", " << val1.x << ", " << val1.y << ", " << val1.z << "]\n"
+        << "Actual: " << expr2 << " [" << val2.w << ", " << val2.x << ", " << val2.y << ", " << val2.z << "]";
+    // clang-format on
+  }
+  testing::AssertionResult expectNearHelper(const char* expr1, const char* expr2, const char* abs_error_expr,
+                                            const std::vector<geometry_msgs::Pose>& val1,
+                                            const std::vector<geometry_msgs::Pose>& val2, double abs_error)
+  {
+    if (val1.size() != val2.size())
+      return testing::AssertionFailure() << "Different vector sizes"
+                                         << "\nExpected: " << expr1 << " (" << val1.size() << ")"
+                                         << "\nActual: " << expr2 << " (" << val2.size() << ")";
+
+    for (size_t i = 0; i < val1.size(); ++i)
+    {
+      ::std::stringstream ss;
+      ss << "[" << i << "].position";
+      GTEST_ASSERT_(isNear((expr1 + ss.str()).c_str(), (expr2 + ss.str()).c_str(), abs_error_expr, val1[i].position,
+                           val2[i].position, abs_error),
+                    GTEST_NONFATAL_FAILURE_);
+
+      ss.str("");
+      ss << "[" << i << "].orientation";
+      GTEST_ASSERT_(isNear((expr1 + ss.str()).c_str(), (expr2 + ss.str()).c_str(), abs_error_expr, val1[i].orientation,
+                           val2[i].orientation, abs_error),
+                    GTEST_NONFATAL_FAILURE_);
+    }
+    return testing::AssertionSuccess();
+  }
+
+  void searchIKCallback(const geometry_msgs::Pose& ik_pose, const std::vector<double>& joint_state,
+                        moveit_msgs::MoveItErrorCodes& error_code)
+  {
+    std::vector<std::string> link_names = { tip_link_ };
+    std::vector<geometry_msgs::Pose> poses;
+    if (!kinematics_solver_->getPositionFK(link_names, joint_state, poses))
+    {
+      error_code.val = error_code.PLANNING_FAILED;
+      return;
+    }
+
+    EXPECT_GT(poses[0].position.z, 0.0f);
+    if (poses[0].position.z > 0.0)
```

Comment:
This check feels somewhat arbitrary, why should it be zero every time?

Reply:
Fully agree. I don't understand this test anyway. However, you are the one, who introduced it ;-)

Reply:
No wonder it looked familiar,  if I recall correctly probably had something to do with the original PR2 testing which inspired this whole unit test program but I'm ok with removing it.